### PR TITLE
[backend] Detect + guard against strategies backtesting the wrong asset

### PR DESCRIPTION
**Pull request by dbrowneup**
## Defect

Confirmed on production data 2026-07-27: strategies silently backtest against the
**wrong price series**.

- `capital_preservation_tbill` declares `ASSET_UNIVERSE = ["BIL"]` (SPDR 1-3 Month
  T-Bill ETF, real-world annualised vol ~0.2%). Its persisted backtest has
  annualised vol **18.65%** and annualised return **10.17%** — that is the S&P, not
  BIL.
- `maillard_2010_risk_parity` declares a 5-asset equal-risk-contribution book
  (`SPY, NIKKEI, GOLD, TREASURY, OIL`), which by construction should have LOWER vol
  than pure equity. Its persisted backtest has vol **18.46%** — also pure equity.
- `pipeline_buy_hold` is 18.65% and that IS correct (it really is SPY).
- Both defective strategies **pass the rigor gate** on returns that are not theirs.

**Root cause** (read in `analytics-engine/src/archimedes_analytics_engine/cli.py::run_command`,
not guessed): a strategy only gets its declared universe fed to it when its class
declares `REQUIRED_FEEDS == 2` (pairs). Every other strategy — single-asset AND
multi-asset alike — is backtested against `RunConfig.operations`
(`BACKTEST_OPERATIONS` env var, default `"SPY"`), **regardless of its own declared
`ASSET_UNIVERSE`**. A strategy with no real timing signal on that default feed
(capital_preservation_tbill), or one whose `self.datas` collapses to the single
default feed (maillard, which needs `engine.run_multi_backtest`), reproduces the
pure buy-and-hold S&P series almost exactly. This also explains why
`analytics-engine/scripts/regen_fixtures.py`'s own docstring says
"capital_preservation_tbill models a T-bill yield, not a TLT buy-hold" — that belief
is itself stale/wrong, corroborating the audit finding.

## What this PR does NOT do (anti-goals, honored)

- Does **not** re-run or regenerate any backtest.
- Does **not** modify any strategy file (`analytics-engine/strategies/**` untouched).
- Does **not** touch the rigor gate or `num_trials`.
- Does **not** touch `strategies_routes.py`, `vaults_routes.py`, `contracts/**`, or
  any of the listed `infra/*` files.

## Deliverable 1 — re-runnable detector: `backend/archimedes/scripts/audit_backtest_universe.py`

Reads the exact store the live rigor gate reads
(`services/backtest_repository.py::get_daily_returns`, keyed by `strategy_id`), and
also reports the delta against the second, independent `strategy_daily_returns`
store (keyed by file `stem`) wherever both exist — that disagreement is itself a
defect signal per the task spec.

Threshold design (documented in full in `services/vol_plausibility.py`'s module
docstring — this is the "defensible rule, honest about its limits" the task asked
for), two complementary rules:

- **Rule A — asset-class plausibility band.** Every declared ticker is resolved to
  a broad vol bucket via the SAME curated classification the on-chain universe
  already uses (`archimedes.universe.SYNTHETIC_UNIVERSE`, `asset_class` per
  `yfinance_ticker`) plus a tiny, explicit supplemental table for the handful of
  legacy `NIKKEI`/`GOLD`/`TREASURY`/`OIL` operation aliases and the `KO`/`PEP`
  pairs legs that predate that SSOT. Each bucket has a wide, generous historical
  vol band (a sanity check, not an oracle — deliberately erring toward
  under-flagging). Catches `capital_preservation_tbill` on its own.
- **Rule B — equity-baseline match.** `pipeline_buy_hold`'s persisted vol is the
  confirmed-correct S&P reference. Any OTHER strategy whose declared universe isn't
  that trivial single-broad-equity book, but whose realized vol lands within a
  tight tolerance of that baseline, is exhibiting the exact statistical signature
  of the single-feed-default fallback — this is what catches
  `maillard_2010_risk_parity` (Rule A's wide multi-asset band alone does **not**
  catch it, by design — verified by a dedicated test asserting Rule-A-only comes
  back clean on that case).

Degenerate (zero-variance) series get their own `degenerate` status — matching
`services/_rigor_helpers.py::compute_dsr`'s exact `np.ptp(arr) == 0.0` check — never
a silent pass, never a crash. Unclassifiable tickers get `unclassified` ("needs
human review" per the task's own guidance, rather than a fabricated verdict).

```
PYTHONPATH=backend python -m archimedes.scripts.audit_backtest_universe
PYTHONPATH=backend python -m archimedes.scripts.audit_backtest_universe --json
```
Exits non-zero if any strategy is `flagged` / `degenerate` / `unclassified`, or
shows a large cross-store delta.

## Deliverable 2 — permanent guard: `backend/archimedes/scripts/run_backtests.py`

`run_backtests.py` is the writer into `backtest_results` (the primary store the
gate reads) that the task's own hint list names, and per
`insert_backtest_if_missing`'s docstring it's one of exactly three writers into
that table. Before persisting each freshly-computed artifact, the script now:

1. Fetches the pre-existing `pipeline_buy_hold` baseline **once per run** (not per
   strategy — one extra query).
2. Runs the SAME Rule A (+ Rule B when the baseline is available) against the
   candidate's own freshly-computed equity curve.
3. On `flagged` / `degenerate`, raises `VolPlausibilityError` **before**
   `insert_backtest_if_missing` is ever called — nothing is written — logs loudly
   at ERROR (no traceback, an explanatory error like the existing `FeedArityError`
   pattern), and the strategy is counted in `failed` with the reason in
   `summary["errors"]`.
4. Degrades gracefully to Rule-A-only the first time ever run (before
   `pipeline_buy_hold` itself has a persisted backtest) — never crashes, never
   blocks the whole batch.

I considered hooking this into `insert_backtest_if_missing` directly (the single
funnel point for all 3 writers) instead, but `strategy_provider.py` already imports
`backtest_repository.py` — adding a provider/universe lookup there would create a
circular import. `run_backtests.py` is the scheduled/cron regeneration entrypoint
and was the only one of the three writers the task named as a candidate location,
so that's where the guard lives. **Deliberately not wired into
`seed_backtests_from_artifacts.py`** (a secondary, artifact-loading-only path for
container-deploy contexts without the analytics-engine installed) or
`analytics-engine/scripts/regen_fixtures.py` (writes to the legacy
`backtest_fixtures.json` bundle, not the `backtest_results` table the gate reads,
per its own docstring — a different store than the one this task specifies) — noting
this honestly rather than silently expanding scope.

## Deliverable 3 — tests (hermetic, no DB/network)

- `backend/tests/test_vol_plausibility.py` (24 tests) — pure unit tests of the
  classification + statistics + two-rule verdict logic, including the
  Rule-A-alone-does-NOT-catch-maillard case, margin-override sensitivity, and the
  exact degenerate/insufficient-data/pathological-return edge cases.
- `backend/tests/test_audit_backtest_universe.py` (7 tests) — sqlite-in-memory +
  temp strategy tree, mocking at the boundary (DB session factory, strategies
  dir), covering: clean strategy passes; BIL-declared 18%-vol strategy flagged
  end-to-end through the CLI (`main()` exit code 1); strategy with no persisted
  returns is skipped, not crashed; degenerate all-zero series reported as its own
  category; maillard-style multi-asset match; cross-store delta reporting.
- `backend/tests/test_run_backtests_vol_guard.py` (5 tests) — same sqlite pattern
  as the existing `test_run_backtests_script.py`: wildly-inconsistent artifact
  refused + zero rows written; loud ERROR log with no traceback; plausible
  artifact persisted normally (no false positive); degenerate artifact refused;
  missing-baseline degrades gracefully.

## Verification

```
$ export PATH="$(conda info --base)/envs/archimedes/bin:$PATH"

$ ruff format --check .
488 files already formatted

$ ruff check --select E9,F63,F7,F40,F82 .
All checks passed!

$ PYTHONPATH=backend python -m pytest backend/tests/test_vol_plausibility.py \
    backend/tests/test_audit_backtest_universe.py \
    backend/tests/test_run_backtests_vol_guard.py \
    backend/tests/test_run_backtests_script.py backend/tests/test_backtest_repository.py -q
40 passed, 1 warning in 2.12s

$ python -m pytest -m "not integration" -q
2975 passed, 2 skipped, 2 deselected, 111 warnings in 190.54s (0:03:10)
```

Full existing suite is green — no regressions from the `run_backtests.py` change
(verified `test_run_backtests_is_idempotent` and
`test_run_backtests_surfaces_typed_errors_without_traceback` still pass unmodified).

## What I deliberately did NOT do

- Did not re-run/regenerate any backtest against real data (no DB access in this
  environment; all verification is against hermetic sqlite fixtures + synthetic
  return series calibrated to the confirmed real numbers).
- Did not wire the guard into `seed_backtests_from_artifacts.py` or
  `analytics-engine/scripts/regen_fixtures.py` — see Deliverable 2 above for why.
- Did not touch `analytics-engine/scripts/regen_fixtures.py`'s stale
  "capital_preservation_tbill models a T-bill yield" comment, even though it
  appears to reflect the same incorrect belief the audit disproves — flagging it
  here for Dan/Önder rather than editing a file outside this task's scope.

Co-Authored-By: Claude Opus 5 (1M context) <noreply@anthropic.com>

### CODE REVIEW

File: backend/archimedes/services/vol_plausibility.py [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+        return VolStats(n_obs=n, annualized_vol=None, annualized_return=None, degenerate=False, insufficient_data=True)
+
+    if float(np.ptp(arr)) == 0.0:
+        return VolStats(n_obs=n, annualized_vol=None, annualized_return=None, degenerate=True, insufficient_data=False)
+
+    sigma = float(arr.std(ddof=1))
+    ann_vol = sigma * math.sqrt(_ANNUALIZATION)
+
+    curve = np.cumprod(1.0 + arr)
+    ann_return: float | None = None if curve[-1] <= 0 else float(curve[-1] ** (_ANNUALIZATION / n) - 1.0)
```
[the remark this answers is not in the document]
Comment (dbrowneup) [reply backend/archimedes/services/vol_plausibility.py]: Correct — fixed in 05ee24d5. `curve = np.cumprod(1.0 + arr)`, so an EVEN number of sub-(-100%) days flips the sign back positive and the tail check misses it. Confirmed against the real code before fixing: `[-1.5, -1.5] + [0.0]*10` produced `annualized_return = -0.9999999999997726` rather than `None`. Now checks `(curve <= 0).any()` over the whole path. Test pins exactly that series and was verified to fail without the fix.

File: backend/archimedes/services/vol_plausibility.py [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+
+    baseline_delta: float | None = None
+    if baseline_vol is not None:
+        baseline_delta = realized_vol - baseline_vol
+        reasons.extend(
+            _evaluate_baseline_match(
+                universe, realized_vol, baseline_vol, abs_tol=baseline_abs_tol, rel_tol=baseline_rel_tol
+            )
+        )
+
```
[the remark this answers is not in the document]
Comment (dbrowneup) [reply backend/archimedes/services/vol_plausibility.py:554]: Correct — fixed in 05ee24d5. Rule A already bailed via `band is None`; Rule B had no equivalent guard, and since `if reasons:` is evaluated before the `not is_fully_classified` branch, the fabricated baseline-match reason won. That contradicts the module docstring's own promise that unclassifiable tickers get `unclassified` rather than a fabricated verdict.

Worth recording the practical consequence, since it is not just cosmetic: the write-time guard in `run_backtests.py` blocks on `flagged`/`degenerate` but not `unclassified`, so this could refuse a legitimate write and blame an equity-baseline match when the real cause was an unrecognised ticker. Now gated on `universe.is_fully_classified`; test asserts `unclassified` (not `flagged`) for `["SPY", "TOTALLY_MADE_UP_TICKER_XYZ"]`, verified to fail without the fix.

File: backend/archimedes/scripts/run_backtests.py [hunk 3]
```diff
@@ -166,6 +236,22 @@ def run_backtests() -> dict:
...
             content_hash = canonical_artifact_hash(artifact_payload)
 
+            # Deliverable-2 permanent guard (audit 2026-07-27): refuse to persist
+            # a backtest whose realized vol is wildly inconsistent with the
+            # strategy's own declared ASSET_UNIVERSE — this is exactly the
+            # confirmed "backtesting the wrong asset" defect class (e.g.
+            # capital_preservation_tbill declaring BIL but reading pure equity
+            # vol). Checked BEFORE insert_backtest_if_missing so a flagged
+            # artifact is never written, loudly, fail-closed.
+            candidate_returns = daily_returns_from_equity_curve(mapped.equity_curve)
```
[the remark this answers is not in the document]
Comment (dbrowneup) [reply backend/archimedes/scripts/run_backtests.py]: Fixed in 05ee24d5 — the guard now pulls the chosen operation's raw `daily_returns` off the artifact payload by name match, falling back to the equity-curve derivation only when the field is absent. See the follow-up thread for why that alone was not sufficient.

File: backend/archimedes/scripts/run_backtests.py [hunk 4]
```diff
@@ -166,6 +265,40 @@ def run_backtests() -> dict:
             )
             content_hash = canonical_artifact_hash(artifact_payload)
```
[the remark this answers is not in the document]
Comment (dbrowneup) [reply backend/archimedes/scripts/run_backtests.py:323]: This one is the important one, and you are right that my first fix only moved the problem — closed properly in a37e9730.

Making the guard read the chosen operation did not help while `get_daily_returns()` still returns the first non-empty `daily_returns` and ignores the `operation` column. With `BACKTEST_OPERATIONS="NIKKEI,SPY"` the guard would validate SPY while the rigor gate and the audit script graded NIKKEI. **A guard that certifies a different series than the gate reads is worse than no guard** — it puts a "checked" stamp on the wrong asset, which is exactly the defect class this PR exists to catch.

Took your suggested shape: persist the artifact with the selected operation moved to the front of `results`, and compute `content_hash` from that persisted payload. "First entry" and "selected operation" are then the same row by construction, so the naive reader is correct without touching it.

I deliberately did **not** fix `get_daily_returns()` to honour the `operation` column here, even though that is the better long-term shape — it is a shared reader, and changing it would re-grade already-persisted rows. That deserves its own PR with its own before/after evidence.

Note it is a no-op under the current default (`BACKTEST_OPERATIONS="SPY"` ⇒ one result): the helper returns the same object when there is nothing to reorder, so no `content_hash` churn and no duplicate re-inserts. It only starts mattering when a second operation is configured — which is exactly when it has to be right. Tests cover reorder, single-op, already-first, unknown-op and `None`.
